server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -4338,6 +4359,28 @@ public void afterRefresh(boolean didRefresh) {
         }
     }
 
+    /**
+     * Switch the current engine to a read-only engine.
+     */
+    public void resetToReadOnlyEngine() throws IOException {
+        assert Thread.holdsLock(mutex) == false : "resetting engine under mutex";
+        assert getActiveOperationsCount() == OPERATIONS_BLOCKED
+            : "resetting engine without blocking operations; active operations are [" + getActiveOperationsCount() + ']';
+        var engineConfig = newEngineConfig(replicationTracker);
+        synchronized (engineMutex) {
+            verifyNotClosed();
+            IOUtils.close(currentEngineReference.get());
+
+            var newEngine = createEngine(engineConfig);
```

Comment:
does this really create a new read only engine? or it depends on some internal state? it's a bit confusing.

Reply:
It depends. Specifically `Stateless#getEngineFactory` may create a regular or readonly engine depending. So maybe lets rename the function to `resetEngine` or `recreateEngine` or something along these lines.

Reply:
I've renamed the method to `resetEngine`